Malware decorator

mattcraddock opened the pull request:
In order to allow for annotating functions passed into the simulation as malware, this pull requests proposes to create a thin wrapper around Callable called LabeledCallable. A LabeledCallable carries a name and a boolean value indicating that it is malware. If we agree that this is the best method for indicating malware I think it will be good to replace this boolean with a more flexible dictionary to allow any properties to be attached. LabeledCallable should also be a subclass of ITObject, which will handle that dictionary in our design. However, since this is a first pass I decided to keep things simple.

The biggest issue with this method is the fact that, in order to maintain backwards compatibility, Callable objects have to be manually checked for the LabeledCallable properties and downcast. In order to minimize the impact of this, I decided to give the Process object the same two properties with defaults so that they are always accessible and no downcasting is needed once the Process is created. This confines downcasting to the constructor for the Process object. If no clients are calling the constructor directly (that is, they are all using the add* methods and creating Processes through the Simulator) then this downcasting can be removed entirely since the add* methods on the Simulator object accept regular Callable objects and pass a LabeledCallable to Process via the generate_for_callable method.

Through the generate_for_callable method, any time a Process adds a new Process to the simulation, the is_malware and label values from the generating Process are passed to the Processes that it spaws, allowing the effects of a specific labeled malware to be passed on throughout the simulation. As above, if we agree on this method it will be necessary to update logging to reflect these new values of the Processes, but for now I decided to keep it simple.

There is also one small tweak to test_random since new versions of pytest do not allow calling approx with strings.

hamelin commented on greensim/__init__.py:
```diff
@@ -442,6 +468,62 @@ def make_happen(*args_event: Any, **kwargs_event: Any) -> None:
     return hook
 
 
+class LabeledCallable(object):
```
I fail to understand why we need this class. I believe the magic of Python closures, the so-called *local parameters* of processes and the delaying of label definition to process instantiation is all we need in order to implement labels. I offer an alternative implementation in this review.

mattcraddock replied:
One more comment here since I'm thinking about how to define the ITObject class in ITSim, which also has labels in our spec: It would be useful to have a generic LabeledObject class in Greensim to standardize the use of labels. In Greensim LabeledCallable could be a subclass of LabeledObject, and in ITSim ITObject could be another subclass of LabeledObject. In that way the storage and access of labels is standardized across any objects in both systems that require labels.

A similar effect could also be achieved by attaching a LabeledObject to a Python function or to local as you described below with the dictionary, so if you have strong opposition to the LabeledCallable concept this may be a useful modification of the dictionary method

hamelin replied:
From my perspective, all callables that should propagate labels are run through `Process` instances. Therefore, would it make sense to write up a `LabeledObject` class, and derive `Process` from it? `Process` already derives from `Named`, as do other basic objects that compose a discrete event simulation. We could modify the name of class `Named` so that it carries more that just a name attribute, and put the label set there. Thus, we leverage my functional construction to put the labels affected to a function into the `Process` instance, and are able to label other objects, as the purpose arises, "for free".

Let's make a final decision on this design aspect in person, and write up the result here.

mattcraddock replied:
I like this solution. Writing this PR already required using Process to manage access to the labels and pass them forward, so it would make sense to have Process own them always. Maybe the @labeled decorator could even produce a Process and remove the need for temporarily storing labels in local?

I also agree with working out the details outside the thread

hamelin commented on greensim/__init__.py:
```diff
@@ -337,6 +337,18 @@ def __init__(self, sim: Simulator, run: Callable, parent: greenlet.greenlet) ->
         self.rsim = weakref.ref(sim)
         self.local = _TreeLocalParam()
         self.local.name = str(uuid4())
+        # This should only be passed LabeledCallable internally
```
Malware denomination in particular seems like a feature for ITsim, in my perspective. However, malware status could rely on process labeling, which is effectively a great greensim addition.

mattcraddock replied:
You're definitely right, I'll move everything specific to malware out to ITsim

hamelin commented on greensim/__init__.py:
```diff
@@ -358,6 +370,20 @@ def resume(self) -> None:
         self.rsim()._schedule(0.0, self.switch)  # type: ignore
 
+    @property
+    def is_malware(self) -> bool:
```
See above -- this is ITsim fare.

hamelin commented on greensim/__init__.py:
```diff
@@ -442,6 +468,62 @@ def make_happen(*args_event: Any, **kwargs_event: Any) -> None:
+            return LabeledCallable(event, str(uuid4()), False)
+
+
+def labeled(label: str, is_malware: bool) -> Callable:
```
This alternative implementation of `labeled` avoids the definition of `LabeledCallable`, leveraging instead the local parameter dictionary all `Process` instances have, as well as closure of the wrapping function around the label name and value.

```
def labeled(label: str, value: bool = True) -> Callable:
    def hook_function(fn_process):
        def wrapper_process(*args, **kwargs):
            local.label = {"name" = label, "value" = value}
            fn_process(*args, **kwargs)
        return wrapper_process
    return hook_function
```

mattcraddock replied:
I might be misunderstanding the description above. fn_process in this case is a Python function and not a Greensim Process, correct? A lot of the logic for these labels could theoretically live in Greensim's Process (which would have all the aspects I argued for above), but since I believe users are meant to define Python functions which Greensim turns into Process objects I focused on how to affect the definition

hamelin replied:
Quite right: the functional construction above only effectively assigns the labels at run time. As we have agreed on Slack, let's finish this review in person.

hamelin commented on greensim/__init__.py:
```diff
@@ -442,6 +468,62 @@ def make_happen(*args_event: Any, **kwargs_event: Any) -> None:
+    return hook
+
+
+def malware(label: str) -> Callable:
```
Why have a name here? From my perspective, I would label a malware process with the construction:

```
@malware
def backdoor():
    ...
```

Then, if I do

```
proc_backdoor = sim.add(backdoor)
```

then I can check something like

```
if proc_backdoor.name_label == "malware" and proc_backdoor.value_label:
   # This process is malware, do something yadda yadda 
```

Why do we want the definition of `malware` to bear a label name, if the label name is "malware" or some such?

Also, the more I think about it, the more the usage of a dictionary for storing label name and value feels rather bad. Instead, we could have a `labels` member to Process defined as a set of strings, and be able to call

```
@labeled("malware", "weird", "madebyme", "hack")
def backdoor():
    ...
```

Then we check:

```
if "malware" in proc_backdoor.labels:
    # It's malware and so on
```

Makes sense?

mattcraddock replied:
I like the solution where labels are treated as a collection of values. I think that's a clean and flexible way to describe the process, though we may want to tweak it for some cases (e.g., if a Process has an owner, do we want a label for each owner or an owner label that takes a value?)

The reason for keeping a name and giving it special treatment was to make sure that the effects of a specific process could be tracked. For example, if there are two malware objects in the simulation with identical labels, even if they pass their labels to everything they touch it will not be possible to tell the effects of one from another unless there is a unique identifier. This doesn't have to be a name specified by the programmer, but I liked that solution since it would allow the programmer to manually set up as many processes as they like under a specific malware by specifying the same name for them all.

hamelin commented on tests/test_random.py:
```diff
@@ -25,7 +25,10 @@ def check_vr(vr, expected, num=5):
+    # New in pytest 3.7
+    if len(expected) > 0 and isinstance(expected[0], float):
+        expected = approx(expected)
+    assert expected == [v for _, v in takewhile(lambda p: p[0] < num, enumerate(vr))]
```
This has already been fixed in the latest version of greensim. I suggest you update your master branch and merge.

mattcraddock replied:
Will do

hamelin commented on tests/test_sim.py:
```diff
@@ -646,3 +646,129 @@ def test_simulator_context_manager(log_destroy):
     assert log_destroy[0:7] == ["A finish", "B EXIT", "B finish", "D EXIT", "D finish", "C EXIT", "C finish"]
+
+
+def test_malware_constructor():
```
Would become useless, should we discard the `LabeledCallable` class.

hamelin commented on tests/test_sim.py:
```diff
@@ -646,3 +646,129 @@ def test_simulator_context_manager(log_destroy):
+
+    run_test_labeled_add(bad_launch, step, True, label)
+
+    @labeled(label, False)
```
Should be a distinct test, I think.

mattcraddock replied:
I'll separate these cases

hamelin commented on tests/test_sim.py:
```diff
@@ -646,3 +646,129 @@ def test_simulator_context_manager(log_destroy):
+
+    run_test_labeled_add(bad_launch, 2 * step, True, label)
+
+    @labeled(label, False)
```
Same.

hamelin commented on tests/test_sim.py:
```diff
@@ -646,3 +646,129 @@ def test_simulator_context_manager(log_destroy):
+
+    run_test_labeled_add(bad_launch, 2 * step, True, label)
+
+    @labeled(label, False)
```
Same.

hamelin commented on tests/test_sim.py:
```diff
@@ -646,3 +646,129 @@ def test_simulator_context_manager(log_destroy):
+        assert guinea_pig.is_malware == Process.current().is_malware
+
+    sim.add(experimentor)
+    sim.run()
```
One test that would be useful is to ensure that when a labeled process `add`s another one, the newly added process (the *addee*?) inherits the labels of the adder.

mattcraddock replied:
This is what lines 680 and 681 in run_test_labeled_add are supposed to do, since last_proc is added by labeled_launcher. However, it isn't obvious and depends on the arguments being set up correctly, so I'll add an explicit test

mattcraddock replied:
This is now covered with the utility function run_test_tagged_add_extra_tag

hamelin commented on tests/test_sim.py:
```diff
@@ -646,3 +646,129 @@ def test_simulator_context_manager(log_destroy):
+
+
+# Case 1: Passed a LabeledCallable
+def test_labeled_callable_generation_copy():
```
Need this if we discard `LabeledCallable`?

mattcraddock replied:
That is correct

mattcraddock replied:
That's correct, these just test that class

hamelin commented on tests/test_sim.py:
```diff
@@ -646,3 +646,129 @@ def test_simulator_context_manager(log_destroy):
+
+
+# Case 2: Process.current() returns TypeError
+def test_labeled_callable_generation_new():
```
Same?

mattcraddock commented:
The latest commits resolve the points discussed above using a new design discussed offline. The TaggedObject class was created to manage tags, which are members of custom Enums inheriting from the root GreensimTag. The LabeledCallable class has been scrapped, and attributes are now stored on the "_greensim_tag_set" element of functions when they are decorated with @tagged. Process is now a subclass of TaggedObject, and upon instantiation a Process will use Process.current() and the tags in its argument to populate its tags. Since this requires special access through the __dict__ object and is not eligible for the typical dot notation the name of the value is stored in the global variable GREENSIM_TAG_ATTRIBUTE

hamelin commented on greensim/tags/__init__.py:
```diff
@@ -0,0 +1,69 @@
+        """
+        return needle in self._tag_set
+
+    def apply(self, new_tag: GreensimTag) -> None:
```
*Bikeshedding warning*

The typical counterpart to `remove`, in most APIs, is `add`. Furthermore, having these super-generic `apply` (or `add`) and `remove` methods to the methods of a class, through subclassing, seems to grab an important part of the namespace. In other words, a subclass of `TaggedObject` might make better, clearer use of method names `apply` and `remove`, given its own application domain.

I suggest the following renames, so as to better manage the subclass namespace:

```
tag_set -> iter_tags (should return an iterator over `_tag_set` -- I would not trust a caller with our own set instance)
apply -> tag (as a verb)
apply_set -> tag_set
remove -> untag
remove_set -> untag_set
clear -> clear_tags
match -> has_tag
```

What do you think?

hamelin commented on greensim/__init__.py:
```diff
@@ -468,62 +462,18 @@ def make_happen(*args_event: Any, **kwargs_event: Any) -> None:
-
-
-def labeled(label: str, is_malware: bool) -> Callable:
+def tagged(tag_set: Iterable[GreensimTag]) -> Callable:
```
A more elegant calling profile would be obtained by defining this as
```
def tagged(*tags: GreensimTag) -> Callable:
```
You thus get to invoke
```
@tagged(Category.VALUE)
def fn():
```
thus losing the double bracketing form, which I find awkward.

hamelin replied:
For giving multiple tags, you'd go
```
@tagged(Category.VALUE, Category.THING, Other.TAG)
def fn():
```

hamelin commented on greensim/tags/__init__.py:
```diff
@@ -0,0 +1,69 @@
+        """
+        self.apply_set([new_tag])
+
+    def apply_set(self, new_tags: Iterable[GreensimTag]) -> None:
```
Instead of having `apply` and `apply_set`, why not have only `apply` and use the variable argument list form to unify both cases? That would net you
```
def apply(self, *tags: GreensimTag) -> None:
```
What you then get is that `tags` is a tuple with all tags you have passed as parameters -- an iterable, in other words.

mattcraddock replied:
Dropped the *_set methods

hamelin commented on greensim/tags/__init__.py:
```diff
@@ -0,0 +1,69 @@
+
+
+@unique
+class GreensimTag(Enum):
```
Mentioning the name of a module in the name of one of its members is awkward. I suggest naming this `TagSpace`, or simply `Enum`. Thus, your customer either imports as is
```
from greensim.tags import Enum

class MyTags(Enum):
    VALUE = "value"
    THING = "thing"
```
or if they already use `enum.Enum` in their own code, they can rename on import:
```
from greensim.tags import Enum as TagEnum

class MyTags(TagEnum):
    ...
```

hamelin replied:
Mind you, the name `TagEnum` would also work. Just not `GreensimTag`. :-D

mattcraddock replied:
You're right, it's a bit verbose. Went with Tags instead

hamelin commented on tests/test_sim.py:
```diff
@@ -7,7 +7,15 @@
+class TestTag(GreensimTag):
+    # Prevent Pytest from complaining
+    __test__ = False
+    ALICE = 0
```
I am sort of surprised this works: it seems that you can thus name-layer the `enum.Enum` class by sub-classing, but if class `GreensimTag` defined a value of its own, then the definition of `TestTag` would raise a `TypeError`.

mattcraddock replied:
Added a comment in code that Tags should remain empty so it can be subclassed

mattcraddock replied:
I also realized that having the `@unique` label does not affect the subclasses, so it is not useful to have it on the empty superclass. I removed it for cleanliness

mattcraddock commented:
Updated the naming and syntax as requested. Also added a test of the tag iterator and dropped the `@unique` label from the newly renamed Tags since it was frivolous
